### ISSUE TITLE
Wordpress KAR resources not deleted due to target cluster getting deleted first

### ISSUE THREAD

**Issue by prasek** [crossplane/app-wordpress#48]
### Problem
`WordpressInstance` KAR resources not cleanly deleted when `WordpressInstance` is deleted.

### Repro
1. Install 
   - crossplane 0.10.0-rc.49.gae92a62
   - provider-azure v0.8.0
   - stack-azure-sample v0.4.0
   - app-wordpress v0.3.0

1. Create `AzureSample` and `WordpressInstance` with ProvisionNewCluster option
1. Delete `WordpressInstance`
1. If the `WordpressInstance` target cluster is deleted before KAR resources, they never get cleanly deleted.

Expected: If the `WordpressInstance` is deleted, then all associated resource including KAR resources and the target cluster are cleanly deleted as well.

```
$ kubectl get kubernetesapplicationresource -n ws1 -o yaml
apiVersion: v1
items:
- apiVersion: workload.crossplane.io/v1alpha1
  kind: KubernetesApplicationResource
  metadata:
    creationTimestamp: "2020-04-14T23:27:15Z"
    deletionGracePeriodSeconds: 0
    deletionTimestamp: "2020-04-14T23:27:16Z"
    finalizers:
    - finalizer.kubernetesapplicationresource.workload.crossplane.io
    generation: 2
    labels:
      app: wp1
    name: wp1-deployment
    namespace: ws1
    ownerReferences:
    - apiVersion: workload.crossplane.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: KubernetesApplication
      name: wp1-app
      uid: 93374721-4ba2-4c34-8118-a7bee3459d39
    resourceVersion: "25893"
    selfLink: /apis/workload.crossplane.io/v1alpha1/namespaces/ws1/kubernetesapplicationresources/wp1-deployment
    uid: 1bf18e80-82d7-40a0-a608-720678b952bc
  spec:
    secrets:
    - name: sql
    targetRef:
      name: wp1-cluster
    template:
      apiVersion: apps/v1
      kind: Deployment
      metadata:
        labels:
          app: wordpress
        name: wordpress
        namespace: wp1
      spec:
        selector:
          matchLabels:
            app: wordpress
        template:
          metadata:
            labels:
              app: wordpress
          spec:
            containers:
            - env:
              - name: WORDPRESS_DB_HOST
                valueFrom:
                  secretKeyRef:
                    key: endpoint
                    name: wp1-deployment-sql
              - name: WORDPRESS_DB_USER
                valueFrom:
                  secretKeyRef:
                    key: username
                    name: wp1-deployment-sql
              - name: WORDPRESS_DB_PASSWORD
                valueFrom:
                  secretKeyRef:
                    key: password
                    name: wp1-deployment-sql
              image: wordpress:4.6.1-apache
              name: wordpress
              ports:
              - containerPort: 80
                name: wordpress
  status:
    conditionedStatus:
      conditions:
      - lastTransitionTime: "2020-04-14T23:27:15Z"
        message: 'cannot create Kubernetes client configuration: cannot get KubernetesTarget
          ws1/wp1-cluster: KubernetesTarget.workload.crossplane.io "wp1-cluster" not
          found'
        reason: Encountered an error during resource reconciliation
        status: "False"
        type: Synced
- apiVersion: workload.crossplane.io/v1alpha1
  kind: KubernetesApplicationResource
  metadata:
    creationTimestamp: "2020-04-14T23:27:15Z"
    deletionGracePeriodSeconds: 0
    deletionTimestamp: "2020-04-14T23:27:16Z"
    finalizers:
    - finalizer.kubernetesapplicationresource.workload.crossplane.io
    generation: 2
    labels:
      app: wp1
    name: wp1-namespace
    namespace: ws1
    ownerReferences:
    - apiVersion: workload.crossplane.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: KubernetesApplication
      name: wp1-app
      uid: 93374721-4ba2-4c34-8118-a7bee3459d39
    resourceVersion: "25894"
    selfLink: /apis/workload.crossplane.io/v1alpha1/namespaces/ws1/kubernetesapplicationresources/wp1-namespace
    uid: cd1639ad-6b97-40c2-9079-b0bfb944588a
  spec:
    targetRef:
      name: wp1-cluster
    template:
      apiVersion: v1
      kind: Namespace
      metadata:
        labels:
          app: wordpress
        name: wp1
  status:
    conditionedStatus:
      conditions:
      - lastTransitionTime: "2020-04-14T23:27:16Z"
        message: 'cannot create Kubernetes client configuration: cannot get KubernetesTarget
          ws1/wp1-cluster: KubernetesTarget.workload.crossplane.io "wp1-cluster" not
          found'
        reason: Encountered an error during resource reconciliation
        status: "False"
        type: Synced
- apiVersion: workload.crossplane.io/v1alpha1
  kind: KubernetesApplicationResource
  metadata:
    creationTimestamp: "2020-04-14T23:27:16Z"
    generation: 1
    labels:
      app: wp1
    name: wp1-service
    namespace: ws1
    ownerReferences:
    - apiVersion: workload.crossplane.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: KubernetesApplication
      name: wp1-app
      uid: 93374721-4ba2-4c34-8118-a7bee3459d39
    resourceVersion: "25897"
    selfLink: /apis/workload.crossplane.io/v1alpha1/namespaces/ws1/kubernetesapplicationresources/wp1-service
    uid: d211226e-bc42-443b-a521-d2f6f6db22f5
  spec:
    targetRef:
      name: wp1-cluster
    template:
      apiVersion: v1
      kind: Service
      metadata:
        labels:
          app: wordpress
        name: wordpress
        namespace: wp1
      spec:
        ports:
        - port: 80
        selector:
          app: wordpress
        type: LoadBalancer
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by hasheddan** [comment]
@prasek I am assuming that the resources were successfully deployed to the cluster before this was observed?

**Comment by prasek** [comment]
@hasheddan yes, everything was working fine

**Comment by prasek** [comment]
@hasheddan came back about 20-30 minutes later and they seem to have been deleted, ideally everything would cleanly delete within seconds of deleting `WordpressInstance`

```
$ kubectl get kubernetesapplicationresource -n ws1 -o yaml
apiVersion: v1
items: []
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by hasheddan** [comment]
Hmm I’ll see if I can repro

**Comment by muvaf** [comment]
I think what happens is that when you delete `WordpressInstance`, both `KubernetesApplication` and `KubernetesCluster` get deleted. If `KubernetesCluster` gets deleted before `KubernetesApplication` or `KubernetesApplicationResource`s, then controllers won't have a chance of sending the deletion signal to remote cluster. Even if they did, it might take a while for their deletion (LB deletion takes some time) and cluster might be down at that point since it's being deleted. 

We might possibly want to introduce a dependency in Crossplane like `block deletion of kubernetes target until all kubernetesapplicationresources targeting it are deleted`

**Comment by hasheddan** [comment]
@muvaf we already intend to delete `KubernetesApplicationResources` if they have deletion timestamp and the `KubernetesTarget` they are scheduled to is not found: 
https://github.com/crossplane/crossplane/blob/84b1c08d18900a99997a02e6417f5b63aeace3f3/pkg/controller/workload/kubernetes/resource/resource.go#L481

I am working on reproducing this and should have a fix in today.

**Comment by hasheddan** [comment]
Ah I know what is happening here. We are are removing the `finalizer`, but then setting requeue to `false` so the status is updated but then we don't ever actually try to delete it. This explains why they were eventually cleaned up because the controller syncs every 1 hour anyway, at which point they would be garbage collected. Fix incoming :+1:

**Comment by muvaf** [comment]
Hmm. Then I think there two problems to fix to make the whole flow correct:
* Claim and its target get deleted immediately and KARs don't go and delete themselves since there is no KubernetesTarget anymore.
* We start deletion of cluster at the same time with KAR deletion. During the process provider might complete the deletion of cluster or shut down the cluster for access before remote resources are deleted. This is especially painful in AWS where load balancer resources do not get deleted unless you specifically delete the Service in the cluster. Since load balancer is still there, VPC deletion gets blocked because of `DependencyViolation`

**Comment by hasheddan** [comment]
For (1), this should be how it currently works and will be fixed if I can reproduce this behavior. For (2), I agree that we need better management of remote resources, but I think that is outside the scope of this issue.

**Comment by prasek** [comment]
just saw this repro with AWS in addition to the Azure case in this ticket, but eventually they deleted, think https://github.com/crossplane/crossplane/pull/1425 ☝️ will fix this, but definitely is happening with more than just stack-azure-stack.

```
apiVersion: workload.crossplane.io/v1alpha1
kind: KubernetesApplicationResource
metadata:
  creationTimestamp: '2020-04-21T23:06:18Z'
  finalizers:
    - finalizer.kubernetesapplicationresource.workload.crossplane.io
  generation: 1
  labels:
    app: wp1
  name: wp1-deployment
  namespace: apr-21-prasek-aws-dev-ws1
  ownerReferences:
    - apiVersion: workload.crossplane.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: KubernetesApplication
      name: wp1-app
      uid: ad2028d8-de1e-46d5-9a37-3a2112b04cb8
  resourceVersion: '752503'
  selfLink: >-
    /apis/workload.crossplane.io/v1alpha1/namespaces/apr-21-prasek-aws-dev-ws1/kubernetesapplicationresources/wp1-deployment
  uid: 934bb7db-c4d4-4949-b33b-e86172f3c863
spec:
  secrets:
    - name: sql
  targetRef:
    name: wp1-cluster
  template:
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      labels:
        app: wordpress
      name: wordpress
      namespace: wp1
    spec:
      selector:
        matchLabels:
          app: wordpress
      template:
        metadata:
          labels:
            app: wordpress
        spec:
          containers:
            - env:
                - name: WORDPRESS_DB_HOST
                  valueFrom:
                    secretKeyRef:
                      key: endpoint
                      name: wp1-deployment-sql
                - name: WORDPRESS_DB_USER
                  valueFrom:
                    secretKeyRef:
                      key: username
                      name: wp1-deployment-sql
                - name: WORDPRESS_DB_PASSWORD
                  valueFrom:
                    secretKeyRef:
                      key: password
                      name: wp1-deployment-sql
              image: 'wordpress:4.6.1-apache'
              name: wordpress
              ports:
                - containerPort: 80
                  name: wordpress
status:
  conditionedStatus:
    conditions:
      - lastTransitionTime: '2020-04-21T23:06:18Z'
        message: >-
          cannot create Kubernetes client configuration: cannot get
          KubernetesTarget apr-21-prasek-aws-dev-ws1/wp1-cluster:
          KubernetesTarget.workload.crossplane.io "wp1-cluster" not found
        reason: Encountered an error during resource reconciliation
        status: 'False'
        type: Synced
```

**Comment by prasek** [comment]
will see if this still repros with v0.10 packages

**Comment by prasek** [comment]
think this issue and the issues below may all be related to the issue of KAR resources not being given the chance to cleanly delete:
- #58 GCP network stuck due to dependencies
- https://github.com/crossplane/provider-aws/issues/201 - AWS VPC stuck due to dependencies
- https://github.com/crossplane/provider-aws/issues/210 - AWS VPC stuck due to dependencies

My suspicion is that in #58 the KAR resources create a `Service` resource which indirectly create side-effect resources (e.g. ELB, FW rules, security groups, etc.) that are in turn preventing the VPC/network from being deleted when the KAR resources aren't given a chance to delete their side-effect resources when they're deleted from the target cluster.

If that's the case, could the KAR resources register finalizers on the `KubernetesApplication` and the `KubernetesApplication` registering finalizers on the `KubernetesTarget` and so on to the `KubernetesCluster` so the cluster doesn't get deleted out from under the KAR resources until they've been given the chance to clean up their side-effect resources?

**Comment by hasheddan** [comment]
@prasek I would think that for #58 deleting the cluster would also clean-up any side-effect resources, but I cannot be certain. I believe I have seen this behavior in cases where I have manually cleaned up the KA's before deleting the cluster and it still blocked network deletion. I think it is more likely that creating a `CloudSQLInstance` in the `Network` is leading to creation of the firewall rules. Will test today and follow up 👍

**Comment by hasheddan** [comment]
After some further investigation, the `CloudSQLInstance` creates a new `VPCNetworkPeering` and augments the `Connection`, but those appeared to be cleaned up successfully. I have found some leaked firewall rules due to creating resources in a `GKECluster` then deleting the cluster.

I believe I have also found the root issue: https://github.com/kubernetes/kubernetes/issues/39420

This is an issue with GCP not cleaning up resources (specifically the network load balancer) on its side, and is pretty difficult for us to work around. Even if we were able to identify orphaned resources and go and delete them for a user when a `Network` is deleted, I am not sure that we should. It is difficult to get into the business of assuming behavior that a user wants. IMO the way to address this situation if we want to clean up orphaned resources for a user is to create a "higher-level" mechanism where a user can ask Crossplane to do a best-effort search and delete for resources that may be blocking. This could require a fair amount of custom logic for each resource and each provider, so I don't know how feasible it is in the short-term.

**Comment by muvaf** [comment]
I think the behavior matches the cloud provider, i.e. if you create a LB in the cluster and delete cluster in the console, the actual LB resource will be leaked in the cloud provider. However, this is an example logic that application controllers usually have during their cleanup phase. In our case, it's a generic controller that we use for Wordpress and it treats all resources equally, i.e. send deletion to `KubernetesCluster`, `KubernetesApplication` and `MySQLInstance` all at once.

IMO, if we want templating-controller to be able to process an app properly, then the devs should be able to specify that logic of ordering.

@hasheddan As we discussed in the sprint planning meeting, I've opened https://github.com/crossplane/templating-controller/issues/17 and will start to work on its implementation.
